### ISSUE TITLE
Thoroughly confused about FMDatabaseQueue and thread safety

### ISSUE THREAD

**Issue by n8stowell82** [ccgus/fmdb#510]
Hi, first off thanks for all the hard work on this wrapper.  I need some help understanding if my implementation of the FMDatabaseQueue is correct.  I feel like I am going crazy.  My application is very data heavy and requires lots of communication with the database.  On top of that, I have a sync process that kicks off on an interval that runs a massive dbsync with a copy of the db file downloaded from my server (multiple device syncing). I recently switched exclusively to FMDatabaseQueue to avoid collisions when accessing the database at the same time the sync could be happening (sync happens on background thread).  I was under the assumption that the queue would prevent these collisions and race conditions when multiple threads are trying to access the same database, and for the most part it has.

Now, here is where I am really confused.  Should I be sharing a single FMDatabaseQueue instance?  I originally tried that but FMDatabase was throwing an error say that wasn't allowed.  Now I just instantiate a new Queue for each attempt to access the database.  this has been working, but randomly I will get a warning like this:
> WARNING: Requested sleep of 87 milliseconds, but SQLite returned 1000. Maybe SQLite wasn't built with HAVE_USLEEP=1?

then I will get an error like this: 
>Unknown error calling sqlite3_step (5: database is locked) eu
or
>-[FMResultSet nextWithError:]:164 Database busy
I am using sqlcipher btw and am for sure calling `setKey` as it is working on other requests.

It seems to fail when trying to update or insert an item into the database, but not every time.

Any help would be greatly appreciated!

**Comment by ccgus** [comment]
You have to share a single FMDatabaseQueue. Any other way, and you're going to have a bad time.

What was the error FMDatabase was giving you?

**Comment by n8stowell82** [comment]
Thanks for the response.  Maybe I am not sharing the Queue correctly? anyway here is the error: 

```assert(currentSyncQueue != self && "inDatabase: was called reentrantly on the same queue, which would lead to a deadlock");```

I assume you cannot call another queue from within a `inDatabase` block (which is what I just did to hit this error) but I have received this error on several other occasions when making many updates in close succession.

To clarify this second point I have a set of methods that wrap all the logic of taking a model and creating a "upsert" query based off of the data in that model and then executing that against the database.  Since each method is self contained they get called one right after another to insert multiple records into the db.  Also I am using SqlCipher  so I have to be able to call `setKey` on each transaction to the db

**Comment by ccgus** [comment]
> I assume you cannot call another queue from within a inDatabase block
You can call another queue- but you can't recursively call inDatabase:. You'll get a lock (print out your stack trace when you get the error- you'll see it).

**Comment by n8stowell82** [comment]
ok. that makes sense.  So to clarify, I **HAVE** to share a single queue throughout the application.  all access on all threads will go through this queue preventing deadlock.  However, I must finish one `inDatabase` before another can be called.

Are these details correct? Thanks again for your help in this matter!

**Comment by n8stowell82** [comment]
Ok sorry for taking more of your time, but I have one last question that I desperately need some help with. According to this thread [https://groups.google.com/forum/#!searchin/sqlcipher/fmdb/sqlcipher/d-j8eRjc7Qg/PpbHDuaBK0IJ](url)  I should not be opening the database every time access to data with sqlCipher as the `setKey` operation takes a significant amount of time.  In the case of using the Queue how would I globally set the key for the database and keep it open for each  subsequent request?

**Comment by n8stowell82** [comment]
more info that has me a bit worried. I updated my app to use a singleton that would distribute a shared copy of FMDatabaseQueue.  I also grouped a bunch of insert commands under a sing `inTransaction` block.  It seems that everything worked fine the first couple of times but then everything just started breaking again.  It starts to fail with an error message of "database is locked" as soon as I call
```
[dbQueue inTransaction:^(FMDatabase *db, BOOL *rollback) { 
  //output in image below starts showing at this point
   [db setKey: [Helper getKey]];
}];
```

<img width="722" alt="dbsync_m" src="https://cloud.githubusercontent.com/assets/1544580/16513713/6cbdac9e-3f2c-11e6-839d-0bcdae4932e2.png">

The only thing I can think of is that I am doing a query in the same `inTransaction` that I am running these inserts, but they are all using the same  `db` variable.  Is there anything wrong doing this?  Thanks again for your help.

**Comment by ccgus** [comment]
If the database is locked, that means another thread or process is hitting it at the same time. Print a backtrace to see if it's another thread.

For setting the key, you probably want to use inDatabase: instead of transaction (you'll want to set the key before doing any db operations I suspect).

**Comment by eltonf** [comment]
I had similar issues to this when I first started using FMDB queues. I suspect you're calling inDatabase inside of [Helper getKey]. You will have to rewrite that method to accept a FMDatabase object. It might look something like this:

`[dbQueue inTransaction:^(FMDatabase *db, BOOL *rollback) { 
  //output in image below starts showing at this point
   [db setKey: [Helper getKeyWithDatabase:db]];
}];`

**Comment by shallad** [comment]
I am sharing a single FMDatabaseQueue throughout my application.  We initialize queue and set the path 
self.dbQueue = [FMDatabaseQueue databaseQueueWithPath:dbPath];

After this initilization , we set the key using:
[self.dbQueue inDatabase:^(FMDatabase *db) {
                [db setKey:"SomeKey"];
                FMResultSet *result = [db executeQuery:@"SELECT count(*) FROM sqlite_master"];
                if ([result next]) {
                    NSLog(@"All good for result: %@" , result);
                }
 }];

I have "n" number of functions/methods which do a DB interaction using inTransaction. 
My question here is : 
Should we set the key in every block of inTransaction or only for first time it is enough

**Comment by eltonf** [comment]
I only set the key one time and that's when I initialize the database connection. Here's a snippet of the code I use:

'''
    _secureDatabase = [FMDatabaseQueue databaseQueueWithPath:filePath];
    [_secureDatabase inDatabase:^(FMDatabase *db) {
        [db setKey:@"SECRET_KEY"];
    }];
'''

**Comment by shallad** [comment]
ok Thank you.. Our team was setting the key in every transaction that we were executing which was significantly hitting the performance.

**Comment by billymeltdown** [comment]
Hi @shallad, as others have noted, you would pay a high performance penalty by closing and re-opening the database, calling setKey on every transaction or query. This and other helpful performance tips for using SQLCipher can be found here:

https://discuss.zetetic.net/t/sqlcipher-performance/14

TLDR; use transactions, normalize your tables, index columns used for searches and joins, and watch out for full table scans (see `cipher_profile` and `EXPLAIN`) :)

**Comment by n8stowell82** [comment]
Thank you all for your comments, it is nice to see I am not the only one.  @eltonf I too have a singleton that distributes the queue, are you saying that immediately after creating the queue for the first time you create an inDatabase call and set key then?  Then having set the key you will not have to do it again for the app lifespan?  Is there a way to force a close on the db as well?

**Comment by eltonf** [comment]
Yes, I create a single instance of the queue, immediately set the key, then never do it again for the lifecycle of the app process. Here is a little more code to give better context.

```objectivec
static FMDatabaseQueue *_secureQueue;

@implementation DataManager

#pragma mark - Initialization

+ (void)initialize
{
    [super initialize];
    
    NSString *filePath = @"PATH_TO_DB_FILE";
    _secureQueue = [FMDatabaseQueue databaseQueueWithPath:filePath];
    [_secureQueue inDatabase:^(FMDatabase *db) {
        [db setKey:@"SECRET_KEY"];
        FMResultSet *rs = [db executeQuery:@"PRAGMA cipher_migrate"]; // Required to migrate database from SQLCipher v2 to v3
        [rs close];
    }];
}

#pragma mark - Data Access

+ (BOOL)exampleDataMethod
{
    __block BOOL success;
    [_secureQueue inDatabase:^(FMDatabase *db) {
       success =  [db executeUpdate:@"INSERT ..."];
    }];
    return success;
}

@end
```

**Comment by n8stowell82** [comment]
Thank you @eltonf I will give this a try.
